### DISCUSSION TITLE
Using Wildcards

### DISCUSSION THREAD

**Discussion by l3112** [open-policy-agent/community#478] [OPA and Rego] [answered]
I have a policy (that works!) -- How would one implement wildcards so that it is read?

`allow[msg]
   bucket := input.aws.s3.buckets[_]
   bucket.name.value == "comp-"_`

Allow buckets if they start with `comp-`
The buckets should always start with comp-, everything after the dash can be whatever.

I did see that ( _ ) is used, and when I tried variations in line 3, the underscore is always considered an error or unsafe except the following;

`bucket.name.value == ["comp"][_]`

and that doesn't give me the results needed.

**Comment by anderseknert** [comment, marked as answer]
Hi @l3112! 

The wildcard can't be used in strings like that. I assume you want to test that **any** bucket name starts with "comp-"? A rule could then use the `startswith` built-in function:

```rego
allow {
    bucket := input.aws.s3.buckets[_]
    startswith(bucket.name.value, "comp-")
}
```
If you're requirement is that **every** bucket name starts with "comp-", you could use the `every` construct, like:

```rego
import future.keywords.every

allow {
    every bucket in input.aws.s3.buckets {
    	startswith(bucket.name.value, "comp-") 
    }
}
```
Obviously might look somewhat different depending on your requirements, but that should be somewhere to start from :)

**Reply by l3112** [reply]
Wonderful, exactly what I needed in conjunction with a `deny` for buckets that don't fit the criteria.

The one last piece is that I want to show where the denied bucket is (which line).

` res := result.new(msg, bucket.name.value)` doesn't work.

**Reply by anderseknert** [reply]
`result.new` is not a function OPA provides, so I don't know what that does. What does the whole rule look like?

**Reply by l3112** [reply]
```
package custom.aws.s3.no_insecure_buckets

import future.keywords.every

allow {
    every bucket in input.aws.s3.buckets {
    	startswith(bucket.name.value, "comp-") #change what's in quotes
    }
}


deny[msg] {
    every bucket in input.aws.s3.buckets {
    	not startswith(bucket.name.value, "comp-") 
    }
    msg := "Bucket name should start with  'comp-'"}`
```

I was looking at the follow code [from here](https://lia.mg/posts/tfsec-and-opa-rego/) as an example;
```
`package custom.aws.s3.no_insecure_buckets

import data.lib.result

deny[res] {
    bucket := input.aws.s3.buckets[_]
    bucket.name.value == "insecure-bucket"
    msg := "Bucket name should not be 'insecure-bucket'"
    res := result.new(msg, bucket.name)
}`
```

**Reply by anderseknert** [reply]
Gotcha! That looks like something specific to tfsec, so unless you're writing policy for that context, that function isn't going to work. You could still return the name of the bucket for each violation:

```rego
deny[msg] {
    bucket := input.aws.s3.buckets[_]
    not startswith(bucket.name.value, "comp-") 
    msg := sprintf("Bucket name %q does not start with 'comp-'", [bucket.name.value])
}
```
If you _are_ writing policy in that context, you'd need to check with the tfsec people how it's meant to work. I have no idea :)

**Reply by l3112** [reply]
Yep, it's tfsec, thank you for all your help!

**Reply by anderseknert** [reply]
Looks like you were missing their import:

```
import data.lib.result
```

Perhaps it was just that?
